### PR TITLE
Serialization: Use a DeclVisitor to implement deserialization safety

### PR DESCRIPTION
**Pull request by tshortli**
Refactor deserialization safety to use a `DeclVisitor` CRTP instead of ad-hoc casts. This ensures every kind of decl is handled explicitly.
    
Resolves rdar://115456536

### CODE REVIEW

**Comment by tshortli** [comment]
@swift-ci please smoke test

File: lib/Serialization/Serialization.cpp [hunk 1]
```diff
@@ -3333,44 +3351,97 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
...
       if (var->isLayoutExposedToClients())
         return true;
 
-      // Consider all lazy var storage as "safe".
+      // Consider all lazy var storage as externally accessible.
       // FIXME: We should keep track of what lazy var is associated to the
-      //        storage for them to preserve the same safeness.
+      //        storage for them to preserve the same accessibility.
       if (var->isLazyStorageProperty())
         return true;
```
Comment (xymus) [inline lib/Serialization/Serialization.cpp:3362]: As the fixme says, this is something that likely marks too many things as safe/accessible. We should keep an eye on it.

File: lib/Serialization/Serialization.cpp [hunk 2]
```diff
@@ -3276,23 +3276,39 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
...
-  /// or otherwise needed by the client in normal builds, this should usually
-  /// correspond to logic in type-checking ensuring these safe decls don't
-  /// refer to implementation details. We have to be careful not to mark
-  /// anything needed by a client as unsafe as the client will reject reading
-  /// it, but at the same time keep the safety checks precise to avoid
-  /// XRef errors and such.
-  static bool isDeserializationSafe(const Decl *decl) {
-    if (auto ext = dyn_cast<ExtensionDecl>(decl)) {
-      // Consider extensions as safe as their extended type.
+  class ExternallyAccessibleDeclVisitor
```
Comment (xymus) [inline lib/Serialization/Serialization.cpp:3279]: I don't mind the name but I think we usually use "exported" for the idea of externally accessible. The term may already be overloaded however.

**Review by xymus — APPROVED** [review]
It's a clear improvement, thanks!

File: lib/Serialization/Serialization.cpp (hunk 2 above)
Comment (tshortli) [reply lib/Serialization/Serialization.cpp:3279]: Yeah, that's a good point. Maybe I should rename this and follow through on extracting this visitor into it's own file the hope that we can unify some or all of the other exportability checks in the compiler.

Comment (xymus) [reply lib/Serialization/Serialization.cpp:3279]: There is a set of `isExported` functions that are also getting very close to this version as you're reducing the difference between was is serialization safe and just what's exported.

Comment (tshortli) [reply lib/Serialization/Serialization.cpp]: I pulled the class out into a header and renamed it to `DeclExportabilityVisitor`.

**Comment by tshortli** [comment]
@swift-ci please smoke test

**Pull request merged**